include/RAJA/pattern/params/forall.hpp

```diff
@@ -0,0 +1,320 @@
+#ifndef FORALL_PARAM_HPP
+#define FORALL_PARAM_HPP
+
+#include "RAJA/policy/sequential/params/reduce.hpp"
+#include "RAJA/policy/tbb/params/reduce.hpp"
+#include "RAJA/policy/openmp/params/reduce.hpp"
+#include "RAJA/policy/cuda/params/reduce.hpp"
+#include "RAJA/policy/cuda/params/kernel_name.hpp"
+#include "RAJA/policy/hip/params/reduce.hpp"
+
+#include "RAJA/util/CombiningAdapter.hpp"
+
+#if defined(RAJA_EXPT_FORALL)
+#define RAJA_EXPT_FORALL_WARN(Msg)
+#else
+#define RAJA_EXPT_FORALL_WARN(Msg) RAJA_DEPRECATE(Msg)
+#endif
+
+namespace RAJA
+{
+namespace expt
+{
+
+  //
+  //
+  // Forall Parameter Packing type
+  //
+  //
+  struct ParamMultiplexer;
+
+  template<typename... Params>
+  struct ForallParamPack {
+
+    friend struct ParamMultiplexer;
+
+    using Base = camp::tuple<Params...>;
```

Comment:
I suppose this has to be a tuple of values to be used on the device. Since this will be a tuple of values we should probably pass it around be reference to avoid unnecessary copies.

Reply:
It also needs to be value as ForallParamPack is thread locally copied in reductions.